WordPress/gutenberg
# Block.json: Refactor and stabilize selectors API

Related:
- https://github.com/WordPress/gutenberg/issues/45194
- https://github.com/WordPress/gutenberg/pull/46579


## What?

Refactors the block.json selectors API.

- Moves the `__experimentalSelector` props within `supports` into their own `selectors` config.
- The block's primary `__experiementalSelector` prop is merged into the new `selectors` config via the `root` prop.
- The selectors API is stabilized as part of this refactor
- Adds greater flexibility allowing individual styles within block support features to be applied to arbitrary selectors
- Allows fallback to appropriate "root" selectors if the individual style isn't set in the selectors config
- Allows entire block support feature to have a single selector applied via a simple string value

## Why?

The experimental selectors have been around for a while now and are beginning to be relied upon more. There's also an increasing need for greater flexibility in setting selectors for individual block support styles.

For more info and history see: https://github.com/WordPress/gutenberg/issues/45194

## How?

#### https://github.com/WordPress/gutenberg/pull/46496/commits/d43c09913e02268a25b2918ca7ddbb41de9216e5

Updates Gutenberg’s block registration to polyfill the selectors prop until it is included in the server-provided definitions. Temporary unit test added for the polyfill.

#### https://github.com/WordPress/gutenberg/pull/46496/commits/43257babd9526a91fa2594a41e64b5d148077082

Adds theme.json unit tests for stylesheet generation using blocks leveraging the new selectors API. The old feature selectors test is maintained and could be deprecated as blocks migrate to the stabilised API.

Due to the updated block metadata, a test block needs to be registered during the phpunit bootstrap so that it is present before WordPress runs and the theme.json block metadata is cached.  
  
A follow-up is currently looking at if we can invalidate the metadata cache and move the test block registrations to their respective tests.


#### https://github.com/WordPress/gutenberg/pull/46496/commits/9d073cf07e412b70014ad41151dcdfe2557591c6


Adds the `selectors` to the block.json schema. The available shape for the selectors API almost matches that of the block supports.

An important difference is that it will accept a `root` selectors which replace the old `__experimentalSelector` for the block as a whole as well as act as generic selector for each feature e.g. color, typography etc. 

#### https://github.com/WordPress/gutenberg/pull/46496/commits/b28fd4a8779bc2579cb9e2823411569cd7a102ac

Updates the Image block.json to use the new selectors API instead of `__experimentalBorder.__experimentalSelector`.

#### https://github.com/WordPress/gutenberg/pull/46496/commits/3edc22b88c2cfd6806e72f53efa66a2326c2d627

Add a global function that can be leveraged to retrieve selectors for specific block types and feature/subfeature.

#### https://github.com/WordPress/gutenberg/pull/46496/commits/c7e307de0610dcc24b0db49a61fbdc31c997bc14

Update theme.json class to leverage the new block CSS selector function.

#### https://github.com/WordPress/gutenberg/pull/46496/commits/1a3e55765e659c87b61867c04f022427d44219b3

Adds a `block_type_metadata_settings` filter to the `lib/compat/wordpress-6.2/blocks.php` file ensuring that the selectors metadata is present in the block type settings. This will be required until the minimum supported WP version is >= 6.2.

#### https://github.com/WordPress/gutenberg/pull/46496/commits/447de33ef309c3f3a936bc40af8422be901eca02

Add documentation for Block Selectors API.

## Testing Instructions

1. Confirm theme.json tests pass: `npm run test:unit:php -- --filter WP_Theme_JSON_Gutenberg_Test`
2. Check new global function tests pass: `npm run test:unit:php -- --filter WP_Get_Block_CSS_Selector_Test`
3. Confirm JS block registration tests pass: `npm run test:unit packages/blocks/src/api/test/registration`
4. In your theme.json add some border styles for the `core/image` block.
5. Load the editor, add an Image block, and confirm those styles are still correctly displayed. 
6. Begin cropping the image and ensure that border styles are still applied.
7. Cancel cropping, save the post, and confirm the border on the frontend.


## Next Steps

If the current approach is accepted we'll need a small patch to core to update the [allowed properties during block registration](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks.php#L369-L385).

```diff
diff --git a/src/wp-includes/blocks.php b/src/wp-includes/blocks.php
index 29e375ab89..cd651808bf 100644
--- a/src/wp-includes/blocks.php
+++ b/src/wp-includes/blocks.php
@@ -378,6 +378,7 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 		'attributes'      => 'attributes',
 		'providesContext' => 'provides_context',
 		'usesContext'     => 'uses_context',
+		'selectors'       => 'selectors',
 		'supports'        => 'supports',
 		'styles'          => 'styles',
 		'variations'      => 'variations',

```

Trac Ticket: https://core.trac.wordpress.org/ticket/57585
GitHub PR: https://github.com/WordPress/wordpress-develop/pull/3942
